### ISSUE TITLE
Request for Mirobot Resources, Information & Experiences

### ISSUE THREAD

**Issue by pglombardo** [rirze/py-mirobot#23]
Apologies for posting this here but this Python library provided me with the most success I've had with the Mirobot.

I have an older Mirobot with the 20201126 firmware and I'm looking for developer resources.  The wlkata discussion forum is dead and finding technical documents is close to impossible.

Browsing through the issues here I just found this bit of invaluable info in #21:

> Unfortunately, I fundamentally refuse to work with WLKata's newer firmware versions ever since they made it closed-source and thereby reneging on their Kickstarter promises.

If I can collect up the resources, I'd like to put together a central place for all things Mirobot including the negatives.

Does the community have a links, resources and/or tips for Mirobot users?

CC: @river-chang

**Comment by rirze** [comment]
I recently followed the confusing trails of forks and found WLKata is maintaining their own version of this library here
https://github.com/mushroom-x/wlkata-mirobot-python/

I can't recommend using their library in principle because I feel pretty discouraged by the whole experience. Namely:

- After spending so many man-hours developing a library, they shipped a bastardized version of my library in their Studio app with no mention of my authorship/license or how to use it. I've had people refer to my library's documentation for how to use their internal python API. 
- In the beginning, they simply forked my library to add some simple fixes. No effort was given to upstream those changes in my library or even communicate how those changes worked. (Surprise! It's related to the internal details of their firmware!) 
- Their fork of my library did not have an issues tracker. They disabled it on purpose, I can only assume. Meanwhile on this project, you can see a rich discussion between users and me about various implementation details about Mirobot and its rover.  (which could not be resolved because surprise, we have no idea how the rover firmware is written!)
- Their mirobot firmware is not precise. Luckily, I had read the source code before they took it down. It was a hack onto a by-then 2 year old reversion of some popular gcode firmware at the time (~~the name is eluding me~~ grbl). They couldn't even bother to use the latest version of the project because I suspect it was too "complicated". Mirobot's custom firmware had one extra file with the custom transformations needed by mirobot's axes. From my measurements, it was not conserving some transformations (an operation followed by its inverse operation does not return it to its original spot). I don't know (or care) if this is a hardware, math, or software issue. 
- Had they simply asked for ownership of the repo, I would've gladly obliged. Along with the pypy package name. I didn't want to fracture any community, however small. But what they were doing, was to strong-arm me into depreciating my repo because theirs was official and had insider info about the firmware updates. 

In general, there's a callousness about open-source software that's exhibited by WLKata here. Starting from a lack of understanding what open source is in principle, to openly reneging on Kickstarter promises about staying open. This is not specific to WLKata by any means. Hardware heads usually think open source is just: fork -> modify -> distribute. And if there's any profit to be made by IP, then they will private all code and make it closed-source. 

Not to mention all the other work I did to try to create avenues for Mirobot discussion including creating the r/mirobot subreddit (after it was shown their sketchy forum site was running some CCP malware). No adoption there. 

Overall, I'm pretty burned by this experience. Don't get me wrong, I definitely learned a lot of package development and software practices: 

- This library got me thinking and empathizing with more restrictive copyright licenses for my projects going forward. 
- I over-engineered the design of this library, which was a useful lesson for me personally. 

In conclusion, I'm probably going to archive this library and attach a forward notice to WLKata's maintained library. I don't have the appetite to deal with this product anymore. WLKata has a lot to learn if they want to adopt open-source. It's not as easy as attaching an adjective to their Kickstarter.

**Comment by pglombardo** [comment]
Thanks @rirze - this information is invaluable.  OSS gets damaged by bad actors.  And their closed firmware is a product killer IMO.  I considered buying another Mirobot until I heard this.

If there's no objections, I'd like to keep this issue open to collect up any other experiences and information from Mirobot users.  There seems to be no other place currently.

**Comment by pglombardo** [comment]
Just an update: I've summarised my experience in [this blog post](https://the0x00.dev/robotics/Dont-Buy-The-WLKata-Mirobot/).  It links here for readers to post their own experiences and resources.

**Comment by rirze** [comment]
> Just an update: I've summarised my experience in [this blog post](https://the0x00.dev/robotics/Dont-Buy-The-WLKata-Mirobot/). It links here for readers to post their own experiences and resources.

Thanks for summarizing those points in your blog post. Honestly though, I would rather the grievances be listed on the mirobot subreddit rather than on some disgruntled python developer's issue tracker. I'm not sure which one is more discoverable, but the mirobot subreddit would look more 'neutral' imo.

**Comment by pglombardo** [comment]
I don't disagree but in any case, I posted the article in both /r/mirobot and /r/robotics.  Hopefully we can save at least a few others  the time and pain.

As for myself, I am thinking to now move on and buy a couple smart servos for the next project.

**Comment by rirze** [comment]
When you wrote that the Mirobot is now $1700, I couldn't believe it. I really, really hope it's an upgraded version of the Kickstarter spec, for which I paid $400 and still doubt the worthiness of the product. 

As for this issue and repo, I'm looking to archive this project. Every few months, I get a person file an issue looking to use this library for their Mirobot and as much as I wish to help them, I'm not able to. When the project gets archived, the issues and PRs will be read-only, unfortunately. 

I encourage all reading this thread to go to other forums to discuss Mirobot related topics.
